### PR TITLE
[New Feature] solarized.el as multi color scheme

### PR DESCRIPTION
**Pull request by conao3**
Hi! I implement a new feature, now solarized.el as multi color scheme.

## Description
The solarized.el has face definitions for *many many* Emacs packages, which is a great job!
I'm sorry, but I don't want to do this again in another location, and I think it's wasteful to maintain the same thing in a different repository.

I compiled only the color definition of solarized.el, so that users could create new color themes by selecting the colors they wanted. This is the actual implementation of what @zk-phi suggests (at following blogs in Japanese).
 - https://qiita.com/zk_phi/items/35097fbb3dde4253d428
 - https://qiita.com/zk_phi/items/ec998eaff72989b05b0a

By using the functions I added, user can create a completely different look while using the consolidated functionality as shown below.

### jellybeans
```
;; inspired vim's jellybeans color-theme
(create-solarized-theme-with-pallet 'light 'solarized-jellybeans-light
  '("#202020" "#ffffff"
    "#ffb964" "#8fbfdc" "#a04040" "#b05080" "#805090" "#fad08a" "#99ad6a" "#8fbfdc"))

(enable-theme 'solarized-jellybeans-light)
```
Capture 2019-05-18 20 36 51
Capture 2019-05-18 20 37 09

### mesa
```
;; inspired emacs's mesa color-theme
(create-solarized-theme-with-pallet 'light 'solarized-mesa-light
  '("#000000" "#faf5ee"
    "#3388dd" "#ac3d1a" "#dd2222" "#8b008b" "#00b7f0" "#1388a2" "#104e8b" "#00688b"))

(enable-theme 'solarized-mesa-light)
```
Capture 2019-05-18 20 40 43
Capture 2019-05-18 20 40 25

### solarized
Of course, we can get solarized in the same way!
```
;; inspired emacs's solarized color-theme
(create-solarized-theme-with-pallet 'light 'solarized-solarized-light
  '("#002b36" "#fdf6e3"
    "#b58900" "#cb4b16" "#dc322f" "#d33682" "#6c71c4" "#268bd2" "#2aa198" "#859900"))

(enable-theme 'solarized-solarized-light)
```
Capture 2019-05-18 20 42 28
Capture 2019-05-18 20 42 44

## Note
As an implementation note, macro, `solarized-with-color-variables`, expansion occurs at function definition time because the macro is inside the function.
However, the macro expects a complete list of color schemes, which is an error.

Therefore, I used eval to ensure that macros are expanded when the function is executed.
This changes the indentation but leaves the differences together for clarity.
When this PR is merged, another PR is sent that will only correct the indentation immediately.

Also, my method doesn't generate a file called `*-theme.el`, so loading with `load-theme` results in an error, and we need to enable it with `enable-theme`.
This point needs to be well documented.

Best,

### CODE REVIEW

**Comment by conao3** [comment]
I added new function, `create-solarized-theme-with-pallet`, like `create-solarized-theme`, simply gives the definition of the face to the symbol that stores the theme symbol (by `deftheme`).
However, the current color changes for some reason...

This should be investigated, but if we enable a different theme, the halfway values are overwritten, so there is no practical problem.

If you turn on some different color themes and get lost, you can find out which ones are active by investigating and overwriting the active custom theme.

Note that `setq` does not work because it uses the `defcustom` variable's `:set` functionality.
```
;;;;;
;; investigate current enable custom themes

custom-enabled-themes
;; => (solarized-planet-dark solarized-tron-dark)

;;;;;
;; overwrite `custom-enabled-themes` and run it's `:set` function
;; to apply specified custom theme(s).

(custom-set-variables '(custom-enabled-themes '(solarized-jellybeans-dark)))
;; => nil

;;;;;
;; now value

custom-enabled-themes
;; => (solarized-jellybeans-dark)
```

**Comment by thomasf** [comment]
this is a great change which I have been thinking about myself.. I want to review it thoroughly though because it changes a lot of base level stuff and I don't have time for it atm. It's one of my priority "free time" github issues to get to though.

**Comment by conao3** [comment]
Thanks @thomasf! How about @bbatsov?

**Comment by bbatsov** [comment]
@conao3 I like the idea in general, but I wonder if we can make it even more generic. I'm a bit worried that we're adding something that's not really specific to Solarized to the theme, not to mention there might be themes like Zenburn that font-lock even more faces. I'll have to think a bit, but I certainly like the idea.

**Comment by bbatsov** [comment]
To be more concrete - I wonder if we can't create a simple package for the palette creation that other themes can leverage internally.

**Comment by alphapapa** [comment]
A few notes:

1.  `pallet` should be `palette`.
2.  The format of `solarized-color-pallet` should probably be an alist or plist.
3.  Why is `solarized-color-pallet` a `defcustom`?  Are users expected to customize it?
4.  >Also, my method doesn't generate a file called *-theme.el, so loading with load-theme results in an error, and we need to enable it with enable-theme.  This point needs to be well documented.

    It needs to be fixed.  :)
5.  > This should be investigated, but if we enable a different theme, the halfway values are overwritten, so there is no practical problem.  If you turn on some different color themes and get lost, you can find out which ones are active by investigating and overwriting the active custom theme.

    I'm not an expert on the theme system or `custom.el`, but note that themes can be applied over other themes and can be disabled, sort of like a stack.  IME it's not a good idea to enable more than one theme or to try to overwrite active themes, as the results are unpredictable.  It's doubly bad when certain themes set non-cosmetic options (i.e. themes are not limited to setting faces).
4.  How does this compare with the `doom-themes` API?  It basically provides an API to more easily define themes, and even has a Solarized Dark variant.  It's already being used to define a variety of themes.  Maybe it would be better to collaborate on that project than, to a certain extent, reinvent the wheel here.

**Comment by bbatsov** [comment]
> How does this compare with the doom-themes API? It basically provides an API to more easily define themes, and even has a Solarized Dark variant. It's already being used to define a variety of themes. Maybe it would be better to collaborate on that project than, to a certain extent, reinvent the wheel here.

I hadn't seen this before, but it's pretty much what I had in mind. Seems the problem is already solved by them indeed.

**Comment by conao3** [comment]
@bbatsov Thanks review! I like this idea too, but I looking for better ideas...

@alphapapa Thanks review! I fixed point 1, 2, 3.
But about point 4, I have no idea to fix...
I think I need to create each file to solve this issue, but I don't know how to create a theme dynamically with the colors and names user specify. Do I create a file in temp directory?

Point 5, I know that, but I hadn't seen it too.
However, the Solarized color theme is one of the most popular Emacs color themes and the face definition of the external package is well maintained.
My suggestion is that by taking the definitions out of the macro, we can create a variety of color themes using Solarized face definition.

**Comment by conao3** [comment]
Sufficient time has passed since the PR was submitted.
My changes will prevent similar work from cluttering up different theme packages.
And this package becomes the de facto standard for color themes, making it easy to change the look of Emacs.

What are your current thoughts on this PR?

**Comment by thomasf** [comment]
I don't understand how to load this theme properly. If load-theme doesn't work this cant be merged..

**Comment by thomasf** [comment]
At the very least I expect this minimal init to work before I can start reviewing this.
Place this in the solarized-emacs repository.

```diff
diff --git a/minimal-init/.gitignore b/minimal-init/.gitignore
new file mode 100644
index 0000000..81244b9
--- /dev/null
+++ b/minimal-init/.gitignore
@@ -0,0 +1 @@
+/elpa/
\ No newline at end of file
diff --git a/minimal-init/init.el b/minimal-init/init.el
new file mode 100644
index 0000000..dfc512f
--- /dev/null
+++ b/minimal-init/init.el
@@ -0,0 +1,42 @@
+
+(let* ((init-dir (file-name-as-directory (file-name-directory load-file-name)))
+       (elisp-dir (expand-file-name ".." init-dir)))
+
+  (add-to-list 'load-path elisp-dir)
+  (setq custom-theme-load-path (list elisp-dir)
+        package-user-dir (expand-file-name "elpa" init-dir)
+        package-archives '(("gnu" . "https://elpa.gnu.org/packages/") ("melpa" . "https://melpa.org/packages/"))))
+
+
+(require 'package)
+(package-initialize)
+
+(defun require-package (package &optional min-version no-refresh)
+  (if (package-installed-p package min-version)
+      t
+    (if (or (assoc package package-archive-contents) no-refresh)
+        (package-install package)
+      (progn
+        (package-refresh-contents)
+        (require-package package min-version t)))))
+
+(mapc 'require-package
+      '(dash flx-ido projectile))
+
+(setq
+ projectile-sort-order 'recently-active
+ projectile-completion-system 'ido
+ projectile-switch-project-action 'projectile-dired
+ projectile-verbose nil)
+
+(ido-mode 1)
+(ido-everywhere 1)
+(flx-ido-mode 1)
+
+(projectile-global-mode)
+
+(require 'solarized)
+
+
+(load-theme 'solarized-dark)
+(load-theme 'solarized-light)
diff --git a/minimal-init/start.sh b/minimal-init/start.sh
new file mode 100755
index 0000000..4dea2bb
--- /dev/null
+++ b/minimal-init/start.sh
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+exec emacs -Q --no-site  -l minimal-init/init.el
```
and start it like this

```sh
./minimal-init/start.sh
```

**Comment by thomasf** [comment]
Always use the solarized-prefix in the code, no `colarized-create-color-palette `

**Comment by conao3** [comment]
Ah, that function name is just typo. If you mention that, you should consider rename `create-solarized-theme`.
Please use [package-lint](https://github.com/purcell/package-lint)
```
 solarize…     1   1 warning         Including "Emacs" in the package summary is usually redundant. (emacs-lisp-package)
 solarize…    10   4 warning         You should include standard keywords: see the variable `finder-known-keywords'. (emacs-lisp-package)
 solariz…    45  11 error           You should depend on (emacs "24.1") if you need `color'. (emacs-lisp-package)
 solariz…   142  26 error           You should depend on (emacs "24.1") if you need `color-name-to-rgb'. (emacs-lisp-package)
 solariz…   143  26 error           You should depend on (emacs "24.1") if you need `color-name-to-rgb'. (emacs-lisp-package)
 solariz…   246     warning         All variables and subroutines might as well have a documentation string (emacs-lisp-checkdoc)
 solariz…   322  11 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solariz…   325   1 error           "create-solarized-theme" doesn't start with package's prefix "solarized". (emacs-lisp-package)
 solariz…   332   1 error           "create-solarized-theme-with-palette" doesn't start with package's prefix "solarized". (emacs-lisp-package)
 solari…   348     warning         Argument ‘color-palette’ should appear (as COLOR-PALETTE) in the doc string (emacs-lisp-checkdoc)
 solari…  1044  26 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  1050  26 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  1056  26 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  1062  26 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  1797   8 error           You should depend on (emacs "24.1") if you need `org-block'. (emacs-lisp-package)
 solari…  1803   8 error           You should depend on (emacs "24.1") if you need `org-date'. (emacs-lisp-package)
 solari…  2387   6 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  2474   5 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
 solari…  2476   3 warning         Closing parens should not be wrapped onto new lines. (emacs-lisp-package)
```

**Comment by conao3** [comment]
I forgot changing reference name when change color palette to alist.
Your test is pass in my environment.

**Comment by thomasf** [comment]
It should support being incorporated in normal theme files.. It would be great if you added an example  in the child-theme-example/themes/ directory.

**Comment by conao3** [comment]
I add create theme file funtion. Now try this feature by adding bottom of your test script.
```
(create-solarized-theme-file 'dark 'solarized-wombat-dark
  '("#2a2a29" "#f6f3e8" "#e5c06d" "#ddaa6f" "#ffb4ac" "#e5786d" "#834c98" "#a4b5e6" "#7ec98f" "#8ac6f2"))

(load-theme 'solarized-wombat-dark)
```

**Comment by thomasf** [comment]
Thats probably a useful feature,, you could maybe consider saving to `custom-theme-directory` instead of a separate themes directory.

**Comment by conao3** [comment]
why? My understanding is that the theme-file is a complete temporary file to deceive load-theme, and because of its characteristics, it should be easy to gitignore.
Also, `custom-theme-directory` does not support user changes to `user-emacs-directory` and it is not desirable to rely on this variable.
For example, in my environment, Emacs uses a separate version of `.emacs.d`, and it works well if there is a simple agreement that each package uses `locate-user-emacs-file`.
But If I use that variable, the file is created in a strange location.
```
user-emacs-directory
;;=> "/home/conao/.emacs.d/local/26.3/"
custom-theme-directory
;;=> "~/.emacs.d/"
locate-user-
```

**Comment by thomasf** [comment]
I'm not sure we need the temp file saving.. It's not important that any theme defined in code can use load-theme.. It's just important that it's possible to create regular redistributable themes that can be packaged..

I have tried to put your init.el wombat theme into it's own file now but I seem to have issues with the child theme feature..  No matter what I set `font-lock-comment-face` to the face does not change.

```el
;;; solarized-light-theme.el --- Solarized Light Theme for Emacs.

;; Copyright (C) 2011-2019 Bozhidar Batsov

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:
;;
;; The the dark solarized theme with the wombat palette
;;
;;; Code:

(require 'solarized)


(deftheme solarized-wombat-dark "The the dark solarized theme with the wombat palette")

(create-solarized-theme-with-palette 'dark 'solarized-wombat-dark
  '("#2a2a29" "#f6f3e8"           ; base03 (02 01 00 0 1 2) base3
    "#e5c06d" "#ddaa6f"           ; yellow orange
    "#ffb4ac" "#e5786d"           ; red    magenta
    "#834c98" "#a4b5e6"           ; violet blue
    "#7ec98f" "#8ac6f2"           ; cyan   green
    )
  (lambda ()
    (custom-theme-set-faces
     theme-name
     `(default ((,class (:foreground ,(solarized-color-blend base03 base3 0.15 2) :background ,base03))))
     `(highlight ((,class (:background ,violet))))
     `(font-lock-builtin-face ((,class (:foreground ,magenta))))
     `(font-lock-constant-face ((,class (:foreground ,blue))))
     `(font-lock-comment-face ((,class (:foreground ,base00))))
     ;; `(font-lock-comment-face ((,class (:foreground "#ff0000"))))
     `(mode-line
       ((,class (:foreground ,base2 :background ,(solarized-color-blend base03 base3 0.85 2)))))
     `(mode-line-inactive
       ((,class (:foreground ,base00 :background ,(solarized-color-blend base03 "black" 0.85 2)))))
     `(mode-line-buffer-id ((,class (:foreground ,base3 :weight bold))))
     `(minibuffer-prompt ((,class (:foreground ,base1)))))))

(provide-theme 'solarized-wombat-dark)



(provide 'solarized-wombat-dark-theme)
;;; solarized-wombat-dark-theme.el ends here

```

**Comment by thomasf** [comment]
it seems like I have to wrap the whole theme creation like this, otherwise it won't override

```el
(let ((custom--inhibit-theme-enable nil))
  (create-solarized-theme-with-palette 'dark 'solarized-wombat-dark
    '("#2a2a29" "#f6f3e8"           ; base03 (02 01 00 0 1 2) base3
      "#e5c06d" "#ddaa6f"           ; yellow orange
      "#ffb4ac" "#e5786d"           ; red    magenta
      "#834c98" "#a4b5e6"           ; violet blue
      "#7ec98f" "#8ac6f2"           ; cyan   green
      )
    (lambda ()
      (custom-theme-set-faces
       theme-name
       `(default ((,class (:foreground ,(solarized-color-blend base03 base3 0.15 2) :background ,base03))))
       `(highlight ((,class (:background ,violet))))
       `(font-lock-builtin-face ((,class (:foreground ,magenta))))
       `(font-lock-constant-face ((,class (:foreground ,blue))))
       ;; `(font-lock-comment-face ((,class (:foreground ,base00))))
       `(font-lock-comment-face ((,class (:foreground ,blue))))
       `(mode-line
         ((,class (:foreground ,base2 :background ,(solarized-color-blend base03 base3 0.85 2)))))
       `(mode-line-inactive
         ((,class (:foreground ,base00 :background ,(solarized-color-blend base03 "black" 0.85 2)))))
       `(mode-line-buffer-id ((,class (:foreground ,base3 :weight bold))))
       `(minibuffer-prompt ((,class (:foreground ,base1)))))))
  )
```

**Comment by thomasf** [comment]
moving the application of the child theme before the regular ones might be workiing.

```

(defun solarized-definition (variant theme-name color-palette &optional childtheme)
  "Create a VARIANT of the theme named THEME-NAME.

When optional argument CHILDTHEME function is supplied it's invoked to further
customize the resulting theme."
;;; Color palette
  (eval
   `(solarized-with-color-variables ',variant ,color-palette
      ,(when childtheme
         `(funcall ',childtheme))
;;; Theme Faces
    (custom-theme-set-faces
     ',theme-name
;;;; Built-in
```

**Comment by thomasf** [comment]
this was a preexisting problem

**Comment by conao3** [comment]
I think my goal and your goal is not shared.
My feature is to allow users to freely define themes while using Solarized functions by selecting some colors.
I don't want the wombat I'm using now to be added to this repository.
In order for me to fix it for myself, I have to issue PR every time, and when there is a duplicate theme file in custom-theme-load-path, the subtle behavior of which file is loaded You need to worry about it.

**Comment by thomasf** [comment]
you don't have to use it,  It's good to have a few themes that makes use of these features included because other people can loook at them for guidance when making their own.

**Comment by conao3** [comment]
Well, OK. But I think if wombat theme put repo for sample, just put in `example` directory like childtheme example.

**Comment by thomasf** [comment]
I like to have it in the package, then it's not forgotten.. Writing tests for emacs themes is such a monumental task that it will never happen in this repo. It's ok bundling some simple color themes. We won't customize them a lot.   I think it's starting to get time to merge now, can't find anything in particular that is a problem right now.

**Comment by conao3** [comment]
Thanks!

**Comment by alphapapa** [comment]
@thomasf @bbatsov @conao3 

This PR was merged without addressing the feedback given about the `doom-themes` package:

In https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497263729 I said:

> How does this compare with the doom-themes API? It basically provides an API to more easily define themes, and even has a Solarized Dark variant. It's already being used to define a variety of themes. Maybe it would be better to collaborate on that project than, to a certain extent, reinvent the wheel here.

Then in https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497278196 @bbatsov said:

> I hadn't seen this before, but it's pretty much what I had in mind. Seems the problem is already solved by them indeed.

Then in https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497499162 @conao3 said:

> Point 5, I know that, but I hadn't seen it too.  However, the Solarized color theme is one of the most popular Emacs color themes and the face definition of the external package is well maintained.  My suggestion is that by taking the definitions out of the macro, we can create a variety of color themes using Solarized face definition.

@conao3's response did not *address* the feedback, it dismissed it.

As @bbatsov and I said, it seems that this issue has been solved by `doom-themes`, and therefore it may not be a good idea to significantly increase the complexity of this package by reimplementing something similar.

Ignoring that issue for 2 and a half months and then saying in https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-520904844, "Sufficient time has passed since the PR was submitted," does not *address the issue.*  It is not assumed that a pull request, especially one like this, should be merged by default.

Why was this PR merged without *addressing* that issue?

File: solarized.el [hunk 1]
```diff
@@ -126,68 +131,129 @@ Related discussion: https://github.com/bbatsov/solarized-emacs/issues/158"
...
+    (blue-l    . "#69B7F0")
+    (cyan-d    . "#00736F")
+    (cyan-l    . "#69CABF")
+    (green-d   . "#546E00")
+    (green-l   . "#B4C342"))
+  "The solarized color palette alist.")
 
 ;;; Setup Start
-(defmacro solarized-with-color-variables (variant &rest body)
+(defmacro solarized-with-color-variables (variant color-palette &rest body)
```
Comment (alphapapa) [inline solarized.el]: @thomasf @bbatsov @conao3 Adding this extra argument has broken the signature of this macro and will break user code that uses it.  For example, I use this macro extensively in my own configuration.

Comment (thomasf) [reply solarized.el]: We have made a stable release so there are no promises broken *yet*.  Sorry for not noticing this discssion, github has decided to stop notifying me properly. I need to go to bed now, will try to read up on these issues again tomorrow.

**Comment by conao3** [comment]
calm down please.

@alphapapa You are an outsider (Not a PR submitter, repo author, or collaborator) and your comments will be fully respected under the CoC's consent, but you cannot force us to reflect your comments. 

I checked doom-theme but it doesn't meet the requirements of my submitted feature.
Can it create a theme from the color palette like below?

```
(create-solarized-theme-file 'dark 'solarized-wombat-dark
  '("#2a2a29" "#f6f3e8" "#e5c06d" "#ddaa6f" "#ffb4ac" "#e5786d" "#834c98" "#a4b5e6" "#7ec98f" "#8ac6f2"))

(load-theme 'solarized-wombat-dark)
```

And `solarized-with-color-variables` is a package internal function and should not depend heavily on such functions. (this function uses one dash, but this function used as an internal function)
I planned to remove this macro and propose a simpler implementation in the next PR.

**Comment by thomasf** [comment]
> @thomasf @bbatsov @conao3
> 
> This PR was merged without addressing the feedback given about the `doom-themes` package:
> 
> In [#330 (comment)](https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497263729) I said:
> 
> > How does this compare with the doom-themes API? It basically provides an API to more easily define themes, and even has a Solarized Dark variant. It's already being used to define a variety of themes. Maybe it would be better to collaborate on that project than, to a certain extent, reinvent the wheel here.
> 
> Then in [#330 (comment)](https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497278196) @bbatsov said:
> 
> > I hadn't seen this before, but it's pretty much what I had in mind. Seems the problem is already solved by them indeed.
> 
> Then in [#330 (comment)](https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-497499162) @conao3 said:
> 
> > Point 5, I know that, but I hadn't seen it too.  However, the Solarized color theme is one of the most popular Emacs color themes and the face definition of the external package is well maintained.  My suggestion is that by taking the definitions out of the macro, we can create a variety of color themes using Solarized face definition.
> 
> @conao3's response did not _address_ the feedback, it dismissed it.
> 
> As @bbatsov and I said, it seems that this issue has been solved by `doom-themes`, and therefore it may not be a good idea to significantly increase the complexity of this package by reimplementing something similar.
> 
> Ignoring that issue for 2 and a half months and then saying in [#330 (comment)](https://github.com/bbatsov/solarized-emacs/pull/330#issuecomment-520904844), "Sufficient time has passed since the PR was submitted," does not _address the issue._ It is not assumed that a pull request, especially one like this, should be merged by default.
> 
> Why was this PR merged without _addressing_ that issue?

It's a simple answer for me. I want to reuse this theme with other palettes. The goal is not to build a framework for building themes, it is to reuse this specific theme.  Sometimes that will require small adjustments.

**Comment by alphapapa** [comment]
> The goal is not to build a framework for building themes, it is to reuse this specific theme.

That is, in fact, building a framework for defining themes.

> It's a simple answer for me. I want to reuse this theme with other palettes. 

This package has 189,000 downloads on MELPA, above the 96th percentile.  If you're going to maintain it, you need to consider more than what you personally want.  What matters most is serving users' interests.  If you want to play with the code, you should do so in your own projects, rather than at users' expense.

**Pull request merged**